File: v1/src/main/java/com/api/v1/domain/entities/Book.java
package com.api.v1.domain.entities;

import com.api.v1.dtos.requests.BookRegistrationRequestDto;
import lombok.Getter;
import org.bson.types.ObjectId;
import org.springframework.data.annotation.Id;
import org.springframework.data.mongodb.core.mapping.Document;
import org.springframework.data.mongodb.core.mapping.Field;

import java.time.ZonedDateTime;

@Document(collection = "v1_books")
@Getter
public class Book {

    @Id
    private ObjectId id = new ObjectId();

    @Field
    private String createdAt = ZonedDateTime.now().toString();

    @Field
    private String isbn;

    @Field
    private String title;

    @Field
    private String subtitle;

    @Field
    private String author;

    @Field
    private String field;

    @Field
    private int numberOfPages;

    @Field
    private int version;

    @Field
    private double price;

    @Field
    private String publisher;

    @Field
    private int publishingYear;

    @Field
    private String archivedAt;

    public Book(
            String isbn,
            String title,
            String subtitle,
            String author,
            String field,
            int numberOfPages,
            int version,
            double price,
            String publisher,
            int publishingYear
    ) {
        this.isbn = isbn;
        this.title = title;
        this.subtitle = subtitle;
        this.author = author;
        this.field = field;
        this.numberOfPages = numberOfPages;
        this.version = version;
        this.price = price;
        this.publisher = publisher;
        this.publishingYear = publishingYear;
    }

    public Book update(BookRegistrationRequestDto request) {
        this.id = new ObjectId();
        this.title = request.title();
        this.subtitle = request.subtitle();
        this.author = request.author();
        this.field = request.field();
        this.numberOfPages = request.numberOfPages();
        this.version = request.version();
        this.price = request.price();
        this.publisher = request.publisher();
        this.publishingYear = request.publishingYear();
        this.archivedAt = null;
        return this;
    }

    public void inactive() {
        this.archivedAt = ZonedDateTime.now().toString();
    }

    public String getFullTitle() {
        if (subtitle.isEmpty()) return title;
        return String.format("%s: %s", title, subtitle);
    }

}

File: kotlin/v2/src/main/kotlin/com/api/v2/book/domain/Book.kt
package com.api.v2.book.domain

import com.api.v2.book.dtos.BookModificationRequestDto
import com.api.v2.book.dtos.BookRegistrationRequestDto
import org.springframework.data.annotation.Id
import org.springframework.data.relational.core.mapping.Table
import java.time.Instant
import java.time.ZoneId
import java.util.UUID

@Table("books")
class Book {

    @Id
    var id: UUID? = null
    lateinit var title: String
    var subtitle: String? = null
    lateinit var isbn: String
    lateinit var author: String
    lateinit var field: String
    lateinit var publisher: String
    var numberOfPages: Int = 1
    var version: Int = 1
    var publishingYear: Int = 0
    var createdAt: Instant = Instant.now()
    var creationZoneId: ZoneId = ZoneId.systemDefault()
    var archivedAt: Instant? = null
    var archivingZoneId: ZoneId? = null

    constructor()

    constructor(requestDto: BookRegistrationRequestDto) {
        this.title = requestDto.title
        this.subtitle = requestDto.subtitle
        this.isbn = requestDto.isbn
        this.author = requestDto.author
        this.field = requestDto.field
        this.publisher = requestDto.publisher
        this.numberOfPages = requestDto.numberOfPages
        this.version = requestDto.version
        this.publishingYear = requestDto.publishingYear
    }

    constructor(
        isbn: String,
        requestDto: BookModificationRequestDto,
        createdAt: Instant,
        creationZoneId: ZoneId
    ) {
        this.isbn = isbn
        this.title = requestDto.title
        this.subtitle = requestDto.subtitle
        this.author = requestDto.author
        this.field = requestDto.field
        this.publisher = requestDto.publisher
        this.numberOfPages = requestDto.numberOfPages
        this.version = requestDto.version
        this.publishingYear = requestDto.publishingYear
        this.createdAt = createdAt
        this.creationZoneId = creationZoneId
    }

    fun fullTitle(): String {
        if (subtitle.isNullOrEmpty()) {
            return title
        }
        return "$title: $subtitle"
    }

    fun archive() {
        archivedAt = Instant.now()
        archivingZoneId = ZoneId.systemDefault()
    }

}